pkg/sql/catalog/descs/collection.go

```diff
@@ -347,39 +345,32 @@ func newMutableSyntheticDescriptorAssertionError(id descpb.ID) error {
 }
 
 // GetAllDescriptors returns all descriptors visible by the transaction,
-// first checking the Collection's cached descriptors for validity if validate
-// is set to true before defaulting to a key-value scan, if necessary.
 func (tc *Collection) GetAllDescriptors(ctx context.Context, txn *kv.Txn) (nstree.Catalog, error) {
 	if err := tc.stored.EnsureAllDescriptors(ctx, txn); err != nil {
 		return nstree.Catalog{}, err
 	}
-	var c nstree.MutableCatalog
-	_ = tc.stored.IterateCachedByID(func(desc catalog.Descriptor) error {
-		c.UpsertDescriptorEntry(desc)
-		return nil
-	})
-	_ = tc.uncommitted.iterateUncommittedByID(func(desc catalog.Descriptor) error {
-		c.UpsertDescriptorEntry(desc)
-		return nil
-	})
-	_ = tc.synthetic.iterateSyntheticByID(func(desc catalog.Descriptor) error {
-		c.UpsertDescriptorEntry(desc)
-		return nil
-	})
-	descs := c.OrderedDescriptors()
+	var ids catalog.DescriptorIDSet
+	for _, iterator := range []func(func(desc catalog.Descriptor) error) error{
+		tc.stored.IterateCachedByID,
+		tc.uncommitted.iterateUncommittedByID,
+		tc.synthetic.iterateSyntheticByID,
+		// TODO(postamar): include temporary descriptors?
+	} {
+		_ = iterator(func(desc catalog.Descriptor) error {
+			ids.Add(desc.GetID())
+			return nil
+		})
+	}
 	flags := tree.CommonLookupFlags{
 		AvoidLeased:    true,
 		IncludeOffline: true,
+		IncludeDropped: true,
 	}
-	if err := tc.finalizeDescriptors(ctx, txn, flags, descs, nil /* validationLevels */); err != nil {
+	// getDescriptorsByID must be used to ensure proper validation hydration etc.
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags, ids.Ordered()...)
+	if err != nil {
 		return nstree.Catalog{}, err
 	}
-	if err := tc.hydrateDescriptors(ctx, txn, flags, descs); err != nil {
-		// If we ran into an error hydrating the types, that means that we
-		// have some sort of corrupted descriptor state. Rather than disable
-		// uses of GetAllDescriptors, just log the error.
-		log.Errorf(ctx, "%s", err.Error())
-	}
```

Comment:
This was rewritten to use the common `getDescriptorsByID` path, ensuring proper validation, hydration, etc. The parts above it are now just to warm the cache and collect the IDs.

It's better this way, previously there could be bugs.